core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -149,118 +160,112 @@ class TransactionStateManager(brokerId: Int,
 
       case Some(log) =>
         val buffer = ByteBuffer.allocate(config.transactionLogLoadBufferSize)
 
-        val loadedTransactions = mutable.Map.empty[String, TransactionMetadata]
-        val removedTransactionalIds = mutable.Set.empty[String]
-
         // loop breaks if leader changes at any time during the load, since getHighWatermark is -1
         var currOffset = log.logStartOffset
-        while (currOffset < highWaterMark
-                && loadingPartitions.contains(topicPartition.partition())
-                && !shuttingDown.get()) {
-          buffer.clear()
-          val fileRecords = log.read(currOffset, config.transactionLogLoadBufferSize, maxOffset = None, minOneMessage = true)
-            .records.asInstanceOf[FileRecords]
-          val bufferRead = fileRecords.readInto(buffer, 0)
-
-          MemoryRecords.readableRecords(bufferRead).batches.asScala.foreach { batch =>
-            for (record <- batch.asScala) {
-              require(record.hasKey, "Transaction state log's key should not be null")
-              TransactionLog.readMessageKey(record.key) match {
-
-                case txnKey: TxnKey =>
-                  // load transaction metadata along with transaction state
-                  val transactionalId: String = txnKey.transactionalId
-                  if (!record.hasValue) {
-                    loadedTransactions.remove(transactionalId)
-                    removedTransactionalIds.add(transactionalId)
-                  } else {
-                    val txnMetadata = TransactionLog.readMessageValue(record.value)
-                    loadedTransactions.put(transactionalId, txnMetadata)
-                    removedTransactionalIds.remove(transactionalId)
-                  }
-
-                case unknownKey =>
-                  // TODO: Metrics
-                  throw new IllegalStateException(s"Unexpected message key $unknownKey while loading offsets and group metadata")
-              }
 
-              currOffset = batch.nextOffset
-            }
-          }
-
-          loadedTransactions.foreach {
-            case (transactionalId, txnMetadata) =>
-              val currentTxnMetadata = addTransaction(transactionalId, txnMetadata)
-              if (!txnMetadata.eq(currentTxnMetadata)) {
-                // treat this as a fatal failure as this should never happen
-                fatal(s"Attempt to load $transactionalId's metadata $txnMetadata failed " +
-                  s"because there is already a different cached transaction metadata $currentTxnMetadata.")
+        try {
+          while (currOffset < highWaterMark
+            && loadingPartitions.contains(topicPartition.partition())
+            && !shuttingDown.get()) {
+            buffer.clear()
+            val fileRecords = log.read(currOffset, config.transactionLogLoadBufferSize, maxOffset = None, minOneMessage = true)
+              .records.asInstanceOf[FileRecords]
```

Comment:
The cast to `FileRecords` is not totally safe. I fixed the same problem in `GroupMetadataManager` recently.